File: Task/Integer-roots/Nim/integer-roots.nim
import bignum

proc root(x: Int; n: int): Int =
  if x < 2: return x
  let n1 = (n - 1).culong
  var c = newInt(1)
  var d = (n1 + x) div n
  var e = (n1 * d + x div d.pow(n1)) div n
  while c != d and c != e:
    c = d
    d = e
    e = (n1 * e + x div e.pow(n1)) div n
  result = if d < e: d else: e


var x: Int
x = newInt(8)
echo "3rd integer root of 8 = ", x.root(3)
x = newInt(9)
echo "3rd integer root of 9 = ", x.root(3)
x = newInt(100).pow(2000) * newInt(2)
echo "First 2001 digits of the square root of 2:"
let s = $x.root(2)
for i in countup(0, s.high, 87): echo s.substr(i, i + 86)

File: Task/Integer-roots/Rust/integer-roots.rs
// [dependencies]
// rug = "1.9"

fn shorten(s: &str, digits: usize) -> String {
    if s.len() <= digits + 3 {
        return String::from(s);
    }
    format!("{}...{}", &s[0..digits/2], &s[s.len()-digits/2..])
}

fn main() {
    use rug::{ops::Pow, Integer};

    let x = Integer::from(8);
    let r = Integer::from(x.root_ref(3));
    println!("Integer cube root of {}: {}", x, r);

    let x = Integer::from(9);
    let r = Integer::from(x.root_ref(3));
    println!("Integer cube root of {}: {}", x, r);

    let mut x = Integer::from(100).pow(2000);
    x *= 2;
    let s = Integer::from(x.root(2)).to_string();
    println!("First {} digits of the square root of 2:\n{}", s.len(), shorten(&s, 70));

    let mut x = Integer::from(100).pow(3000);
    x *= 2;
    let s = Integer::from(x.root(3)).to_string();
    println!("First {} digits of the cube root of 2:\n{}", s.len(), shorten(&s, 70));
}
